### PR TITLE
SystemZ: Fold copy of vector immediate to gr128

### PR DESCRIPTION
**Pull request by arsenm**
If materializing a constant in a vector register that is just
going to be copied to general registers, directly materialize
the immediate in the gpr. This will avoid a few lit test regressions
in a future commit.

### CODE REVIEW

**Review by uweigand — CHANGES_REQUESTED** [review]
This doesn't appear to handle the VGBM semantics correctly, see inline comments.

File: llvm/lib/Target/SystemZ/SystemZInstrInfo.cpp [hunk 1]
```diff
@@ -640,6 +640,48 @@ bool SystemZInstrInfo::foldImmediate(MachineInstr &UseMI, MachineInstr &DefMI,
                                      Register Reg,
                                      MachineRegisterInfo *MRI) const {
   unsigned DefOpc = DefMI.getOpcode();
+
+  if (DefOpc == SystemZ::VGBM) {
+    // Fold gr128 = COPY (vr128 VGBM imm)
+    //
+    // %tmp:gr64 = LGHI 0
+    // to  gr128 = REG_SEQUENCE %tmp, %tmp
```
[the remark this answers is not in the document]
Comment (arsenm) [reply llvm/lib/Target/SystemZ/SystemZInstrInfo.cpp:652]: Changed to only handle 0. Generalizing this is left as an exercise for whoever wants to pick up https://github.com/arsenm/llvm-project/tree/systemz-legal-v2i64

File: llvm/test/CodeGen/SystemZ/fold-copy-vector-immediate.mir [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+    ; CHECK-NEXT: {{  $}}
+    ; CHECK-NEXT: [[COPY:%[0-9]+]]:gr64bit = COPY $r2d
+    ; CHECK-NEXT: [[COPY1:%[0-9]+]]:addr64bit = COPY [[COPY]]
+    ; CHECK-NEXT: [[VGBM:%[0-9]+]]:vr128bit = VGBM 1
+    ; CHECK-NEXT: [[COPY2:%[0-9]+]]:gr128bit = COPY [[VGBM]]
+    ; CHECK-NEXT: $r0q = COPY [[COPY2]]
+    ; CHECK-NEXT: Return implicit $r0q
+    %0:gr64bit = COPY $r2d
+    %1:addr64bit = COPY %0
+    %2:vr128bit = VGBM 1
```
Comment (uweigand) [inline llvm/test/CodeGen/SystemZ/fold-copy-vector-immediate.mir:136]: I assume this should be VGBM 0 here as well?

**Review by uweigand — APPROVED** [review]
See inline comment, otherwise LGTM.

File: llvm/test/CodeGen/SystemZ/fold-copy-vector-immediate.mir (hunk 2 above)
Comment (arsenm) [reply llvm/test/CodeGen/SystemZ/fold-copy-vector-immediate.mir:136]: This is supposed to be the non-0 value test

Comment (uweigand) [reply llvm/test/CodeGen/SystemZ/fold-copy-vector-immediate.mir:136]: Ah, I missed that.  That's fine then.

**Comment by vitalybuka** [comment]
Fails here:
https://lab.llvm.org/buildbot/#/builders/239/builds/6893
https://lab.llvm.org/buildbot/#/builders/5/builds/43113
https://lab.llvm.org/buildbot/#/builders/168/builds/20228

**Pull request merged**